### PR TITLE
Avoid concurrently scheduling tied ops with hazards.

### LINKED ISSUE #14889: [Stream] ScheduleConcurrency incorrectly partitions conflicting read/write together
### What happened?

In a first dispatch `%arg0` (`%arg2`) is a read-only resource and produces new value, in second dispatch it's tied to output. ScheduleConcurrency incorrectly places them into the same concurrent region.

```mlir
module {
  func.func @partitioningForConcurrency(%arg0: !stream.resource<external>) -> (!stream.resource<external>, !stream.resource<external>) {
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %c4 = arith.constant 4 : index
    %results:2, %result_timepoint = stream.async.execute with(%arg0 as %arg1: !stream.resource<external>{%c4}) -> (!stream.resource<external>{%c4}, %arg0{%c4}) {
      %1 = stream.async.dispatch @ex::@dispatch_0[%c1, %c1, %c1](%arg1[%c0 to %c4 for %c4]) : (!stream.resource<external>{%c4}) -> !stream.resource<external>{%c4}
      %2 = stream.async.dispatch @ex::@dispatch_1[%c1, %c1, %c1](%arg1[%c0 to %c4 for %c4]) : (!stream.resource<external>{%c4}) -> %arg1{%c4}
      stream.yield %1, %2 : !stream.resource<external>{%c4}, !stream.resource<external>{%c4}
    } => !stream.timepoint
    %0:2 = stream.timepoint.await %result_timepoint => %results#0, %results#1 : !stream.resource<external>{%c4}, !stream.resource<external>{%c4}
    return %0#0, %0#1 : !stream.resource<external>, !stream.resource<external>
  }
}
```

```
build/tools/iree-opt <file> -pass-pipeline="builtin.module(func.func(iree-stream-schedule-concurrency))"
```

```
%1:2 = stream.async.concurrent with(%arg1 as %arg2: !stream.resource<external>{%c4}) -> (!stream.resource<external>{%c4}, !stream.resource<external>{%c4}) {
        %2 = stream.async.dispatch @ex::@dispatch_0[%c1, %c1, %c1](%arg2[%c0 to %c4 for %c4]) : (!stream.resource<external>{%c4}) -> !stream.resource<external>{%c4}
        %3 = stream.async.dispatch @ex::@dispatch_1[%c1, %c1, %c1](%arg2[%c0 to %c4 for %c4]) : (!stream.resource<external>{%c4}) -> %arg2{%c4}
        stream.yield %2, %3 : !stream.resource<external>{%c4}, !stream.resource<external>{%c4}
}
```

`dispatch_0` reads from `%arg2` and `dispatch_1` write into the same resource in the same concurrent region.

Should concurrency partitioning also check all users of all resource arguments here: https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Dialect/Stream/Analysis/Partitioning/ReferencePartitioning.cpp#L359-L362 ?

### PR DESCRIPTION
**Pull request by benvanik**
Fixes #14889.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Stream/Analysis/Partitioning/ReferencePartitioning.cpp [hunk 1]
```diff
@@ -386,6 +386,49 @@ partitionRegionConcurrencyReference(IREE::Stream::PartitioningConfigAttr config,
       // Always inherit hazards whether merging or not.
       opInfo.hazards |= userInfo.hazards;
     }
+
+    // Additional exhaustive testing for users of tied operands.
+    // For each resource operand of this op we scan back to see if there are any
```
Comment (ezhulenev) [inline compiler/src/iree/compiler/Dialect/Stream/Analysis/Partitioning/ReferencePartitioning.cpp]: Not really "scan back", more like "scan forward"? When I read "back" I imagine that we'll traverse all users defined before the `op`

**Review by ezhulenev — APPROVED** [review]
I think there is still a problem with identifying insertion point for formed concurrent region, I'll prepare one more test example.

**Review by ezhulenev — APPROVED** [review]

**Comment by ezhulenev** [comment]
```
func.func @partitioningForConcurrency(%arg0: !stream.resource<external>, 
                                      %arg1: !stream.resource<external>)
  -> (!stream.resource<external>, !stream.resource<external>,
      !stream.resource<external>, !stream.resource<external>) {

  %c0 = arith.constant 0 : index
  %c1 = arith.constant 1 : index
  %c4 = arith.constant 4 : index

  // Wave #0 (reads from %arg0 and %arg1)
  %0 = stream.async.dispatch @ex::@dispatch_0[%c1, %c1, %c1](%arg0[%c0 to %c4 for %c4])
      : (!stream.resource<external>{%c4}) -> !stream.resource<external>{%c4}
  %1 = stream.async.dispatch @ex::@dispatch_0[%c1, %c1, %c1](%arg1[%c0 to %c4 for %c4])
      : (!stream.resource<external>{%c4}) -> !stream.resource<external>{%c4}

  // Wave #1 (writes into %arg0 and %arg1 via tied operands)
  %2 = stream.async.dispatch @ex::@dispatch_1[%c1, %c1, %c1](%arg0[%c0 to %c4 for %c4])
      : (!stream.resource<external>{%c4}) -> %arg0{%c4}
  %3 = stream.async.dispatch @ex::@dispatch_1[%c1, %c1, %c1](%arg1[%c0 to %c4 for %c4])
      : (!stream.resource<external>{%c4}) -> %arg1{%c4}

  return %0, %1, %2, %3 : !stream.resource<external>, !stream.resource<external>,
                          !stream.resource<external>, !stream.resource<external>
}
```

Forms two concurrent regions, and because second one uses only block arguments it gets reordered with the first one:

```
      %1:2 = stream.async.concurrent with(%arg2 as %arg4: !stream.resource<external>{%c4}, %arg3 as %arg5: !stream.resource<external>{%c4}) -> (!stream.resource<external>{%c4}, !stream.resource<external>{%c4}) {
        %3 = stream.async.dispatch @ex::@dispatch_1[%c1, %c1, %c1](%arg4[%c0 to %c4 for %c4]) : (!stream.resource<external>{%c4}) -> %arg4{%c4}
        %4 = stream.async.dispatch @ex::@dispatch_1[%c1, %c1, %c1](%arg5[%c0 to %c4 for %c4]) : (!stream.resource<external>{%c4}) -> %arg5{%c4}
        stream.yield %3, %4 : !stream.resource<external>{%c4}, !stream.resource<external>{%c4}
      }
      %2:2 = stream.async.concurrent with(%arg2 as %arg4: !stream.resource<external>{%c4}, %arg3 as %arg5: !stream.resource<external>{%c4}) -> (!stream.resource<external>{%c4}, !stream.resource<external>{%c4}) {
        %3 = stream.async.dispatch @ex::@dispatch_0[%c1, %c1, %c1](%arg4[%c0 to %c4 for %c4]) : (!stream.resource<external>{%c4}) -> !stream.resource<external>{%c4}
        %4 = stream.async.dispatch @ex::@dispatch_0[%c1, %c1, %c1](%arg5[%c0 to %c4 for %c4]) : (!stream.resource<external>{%c4}) -> !stream.resource<external>{%c4}
        stream.yield %3, %4 : !stream.resource<external>{%c4}, !stream.resource<external>{%c4}
      }
```

File: compiler/src/iree/compiler/Dialect/Stream/Analysis/Partitioning/ReferencePartitioning.cpp (hunk 1 above)
Comment (benvanik) [reply compiler/src/iree/compiler/Dialect/Stream/Analysis/Partitioning/ReferencePartitioning.cpp]: ah yeah, waves are created in reverse, so it's back into the waves that have been defined - will clarify

**Comment by ezhulenev** [comment]
It's a problem in WavePartitionBuilder insertion point logic, and `mlir::sortTopologically` also makes me nervous, I suspect it can also reorder reads/writes.

**Comment by benvanik** [comment]
Yeah that sort deeply bothers me but I haven't found a way to remove it yet. Really would like to have this all rewritten - it was never meant to last this long ;(

**Comment by ezhulenev** [comment]
I'm running tests with https://github.com/openxla/iree/pull/14857 (seems to be obviously correct change, I doubt it has bugs), and I keep getting slightly different results, these are two issues that I was able to spot, but I suspect there are more problems.

**Comment by benvanik** [comment]
What are you generating these programs with? Vulkan, Metal, and CPU all rely on concurrency scheduling and so far we haven't had any issues - wondering if this is being triggered by some path we haven't hit in the normal compiler flow?

**Comment by benvanik** [comment]
Yeah, your posted example here is an issue with the topological sort; the partitions look correct:
```
PARTITION[0]:
 INS:
  %c1, %arg2, %c4, %c0, %arg3
 OUTS:
  %0, %1
 OPS:
  %0 = stream.async.dispatch @ex::@dispatch_0[%c1, %c1, %c1](%arg2[%c0 to %c4 for %c4]) : (!stream.resource<external>{%c4}) -> !stream.resource<external>{%c4}
  %1 = stream.async.dispatch @ex::@dispatch_0[%c1, %c1, %c1](%arg3[%c0 to %c4 for %c4]) : (!stream.resource<external>{%c4}) -> !stream.resource<external>{%c4}
PARTITION[1]:
 INS:
  %c1, %arg2, %c4, %c0, %arg3
 OUTS:
  %2, %3
 OPS:
  %2 = stream.async.dispatch @ex::@dispatch_1[%c1, %c1, %c1](%arg2[%c0 to %c4 for %c4]) : (!stream.resource<external>{%c4}) -> %arg2{%c4}
  %3 = stream.async.dispatch @ex::@dispatch_1[%c1, %c1, %c1](%arg3[%c0 to %c4 for %c4]) : (!stream.resource<external>{%c4}) -> %arg3{%c4}
```

This PR improves things, but we'll need another issue for tracking the sort issue.

**Comment by ezhulenev** [comment]
Maybe, I suspect I might accidentally add dispatches with aliasing buffers that IREE doesn’t recognize as aliasing, working on checking this.

I use tied operants A LOT (everything is tied basically), I think it might not have that much test coverage.

**Comment by benvanik** [comment]
Yeah, I suspect you're doing more than the compiler has done so we've been able to skate by with these issues. Will be good to fix them and get more tests in place as it'd be nice if the compiler could also produce dispatches like this that worked. I think the big lift is going to be around the sorting logic which has been a known issue for ages (including just the other week). Suggestions welcome if you spot anything :) I'm thinking of a topological sort that uses the hazard analysis that partitioning uses but haven't looked at this code in 2 years so there may be something simpler (or ideally, being able to remove sorting entirely).

**Comment by ezhulenev** [comment]
I fixed it by picking insertion point (https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Dialect/Stream/Transforms/ScheduleConcurrency.cpp#L53) right before the first op in the partition (non cloned op?), and removing topological sorting (although the problem was in insertion point and not topo-sorting itself). But I'm not sure it's the correct fix for this problem.

**Pull request merged**